deployment/keystone/changeset/update_don_test.go

```diff
@@ -0,0 +1,165 @@
+	"github.com/smartcontractkit/chainlink/deployment/keystone/changeset"
+	"github.com/smartcontractkit/chainlink/deployment/keystone/changeset/internal"
+	kcr "github.com/smartcontractkit/chainlink/v2/core/gethwrappers/keystone/generated/capabilities_registry"
+	"github.com/smartcontractkit/chainlink/v2/core/services/keystore/keys/p2pkey"
+)
+
+func TestUpdateDon(t *testing.T) {
+	t.Parallel()
+
+	var (
+		capA = kcr.CapabilitiesRegistryCapability{
+			LabelledName: "capA",
+			Version:      "0.4.2",
+		}
+		capB = kcr.CapabilitiesRegistryCapability{
+			LabelledName: "capB",
+			Version:      "3.16.0",
+		}
+		caps = []kcr.CapabilitiesRegistryCapability{capA, capB}
+	)
+	t.Run("no mcms", func(t *testing.T) {
+		te := SetupTestEnv(t, TestConfig{
+			WFDonConfig:     DonConfig{N: 4},
+			AssetDonConfig:  DonConfig{N: 4},
+			WriterDonConfig: DonConfig{N: 4},
+			NumChains:       1,
+		})
+
+		// contract set is already deployed with capabilities
+		// we have to keep track of the existing capabilities to add to the new ones
+		var p2pIDs []p2pkey.PeerID
+		newCapabilities := make(map[p2pkey.PeerID][]kcr.CapabilitiesRegistryCapability)
+		for id, _ := range te.WFNodes {
+			k, err := p2pkey.MakePeerID(id)
+			require.NoError(t, err)
+			p2pIDs = append(p2pIDs, k)
+			newCapabilities[k] = caps
+		}
+
+		t.Run("succeeds if update sets new and existing capabilities", func(t *testing.T) {
+			cfg := changeset.UpdateDonRequest{
+				RegistryChainSel: te.RegistrySelector,
+				P2PIDs:           p2pIDs,
+				CapabilityConfigs: []changeset.CapabilityConfig{
+					{
+						Capability: capA,
+					},
+					{
+						Capability: capB,
+					},
+				},
+			}
+
+			csOut, err := changeset.UpdateDon(te.Env, &cfg)
+			require.NoError(t, err)
+			require.Len(t, csOut.Proposals, 0)
+			require.Nil(t, csOut.AddressBook)
+
+			assertDonContainsCapabilities(t, te.ContractSets()[te.RegistrySelector].CapabilitiesRegistry, caps, p2pIDs)
+		})
+	})
+	t.Run("with mcms", func(t *testing.T) {
+		te := SetupTestEnv(t, TestConfig{
+			WFDonConfig:     DonConfig{N: 4},
+			AssetDonConfig:  DonConfig{N: 4},
+			WriterDonConfig: DonConfig{N: 4},
+			NumChains:       1,
+			UseMCMS:         true,
+		})
+
+		// contract set is already deployed with capabilities
+		// we have to keep track of the existing capabilities to add to the new ones
+		var p2pIDs []p2pkey.PeerID
+		for id, _ := range te.WFNodes {
+			k, err := p2pkey.MakePeerID(id)
+			require.NoError(t, err)
+			p2pIDs = append(p2pIDs, k)
+		}
+
+		cfg := changeset.UpdateDonRequest{
+			RegistryChainSel: te.RegistrySelector,
+			P2PIDs:           p2pIDs,
+			CapabilityConfigs: []changeset.CapabilityConfig{
+				{
+					Capability: capA,
+				},
+				{
+					Capability: capB,
+				},
+			},
+			MCMSConfig: &changeset.MCMSConfig{MinDuration: 0},
+		}
+
+		csOut, err := changeset.UpdateDon(te.Env, &cfg)
+		require.NoError(t, err)
+
+		if true {
```

Comment:
/nit probably on purpose, but trivially always evaluates here

Reply:
ah, yes. thanks that is crufty debugging. 

will rm in quick follow up